powersimdata/input/converter/pypsa_to_grid.py

```diff
@@ -286,13 +282,77 @@ def _read_network(self, n, add_pypsa_cols=True):
         dcline["from_bus_id"] = pd.to_numeric(dcline.from_bus_id, errors="ignore")
         dcline["to_bus_id"] = pd.to_numeric(dcline.to_bus_id, errors="ignore")
 
-        # storage
-        storage_gen_storageunits = _get_storage_gen(n, "storage_units")
-        storage_gencost_storageunits = _get_storage_gencost(n, "storage_units")
-        storage_storagedata_storageunits = _get_storage_storagedata(n, "storage_units")
-        storage_gen_stores = _get_storage_gen(n, "stores")
-        storage_gencost_stores = _get_storage_gencost(n, "stores")
-        storage_storagedata_stores = _get_storage_storagedata(n, "stores")
+        # storage units
+        c = "storage_units"
+        storage_gen_storageunits = _get_storage_gen(n.storage_units, c)
+        storage_gencost_storageunits = _get_storage_gencost(n.storage_units, c)
+        storage_storagedata_storageunits = _get_storage_storagedata(n.storage_units, c)
+
+        inflow = n.get_switchable_as_dense("StorageUnit", "inflow")
+        has_inflow = inflow.any()
+        if has_inflow.any():
+            # add artificial buses
+            suffix = " inflow"
+
+            def add_suffix(s):
+                return str(s) + suffix
+
+            storage_gen_inflow = storage_gen_storageunits[has_inflow]
+            buses_old = storage_gen_inflow.bus_id
+            buses_new = buses_old.astype(str) + suffix
+            bus_inflow = bus.reindex(buses_old).rename(index=add_suffix)
+
+            # add discharging dcline (has same index as inflow storages)
+            dcline_inflow = pd.DataFrame(
+                {
+                    "from_bus_id": buses_new,
+                    "to_bus_id": buses_old,
+                    "Pmax": storage_gen_inflow.Pmax,
+                }
+            )
+            dcline_inflow.index += " discharge"
+
+            # add charging dcline (has same index as inflow storages)
+            dcline_inflow = pd.DataFrame(
+                {
+                    "from_bus_id": buses_old,
+                    "to_bus_id": buses_new,
+                    "Pmax": -storage_gen_inflow.Pmin,
+                }
+            )
+            dcline_inflow.index += " charge"
+
+            # add inflow generator
+            gen_inflow = storage_gen_inflow.rename(index=add_suffix)
+            gen_inflow["Pmax"] = n.storage_units_t.inflow.max().rename(add_suffix)
```

Comment:
Not related to this PR: @rouille `Pmin` configuration of these inflow generators matters here, if we do `Pmin = Pmax = profile value at the hour`, certain corner case could give us infeasibility that can't be resolved by the engine: the coupled storage device (reservoir) is at full energy level and profile value at the hour is greater than the discharging capacity (`Pmax` of the dcline). In other words, we will need to set `Pmin = 0` for these inflow generators to handle the spillage.